### ISSUE TITLE
Plugins: Move IHostServiceProxy and IServiceProxy to Coral.PluginBase

### ISSUE THREAD

**Issue by CaffeinatedCoder** [GOATS2K/Coral#101]
While studying your code base I found a potential coupling issue: you are declaring both types IHostServiceProxy and IServiceProxy as well as their implementations in the same assembly, namely in Coral.PluginHost. This introduces a potential problem: every plugin assembly so now needs to reference Coral.PluginHost for the sole purpose of gaining access to one of the types IHostServiceProxy or IServiceProxy.
In the same time all other types in Coral.PluginHost are now accessible to every plugin as well, even those that are solely related to hosting plugins in general, not about implementing them. 
To my understanding this is an architectural segregation issue. Plugin assemblies should never care about the way by whom or how they are going to be hosted.

I would propose to move IServiceProxy and IHostServiceProxy to Coral.PluginBase and make plugins depend only on Coral.PluginBase. This would ensure that plugins are given access only to the most essential interfaces and types they need to provide a working plugin implementation.

**Comment by GOATS2K** [comment]
Hi!

Thank you for your comment - how did you find the project?

`IHostServiceProxy` allows plugins to gain access to certain services provided by the host. The type is then registered with the plugin's service provider so it can be accessed by the plugin. It resides within the `Coral.PluginHost` assembly because I wanted keep everything related to communication between the host and the plugin in one project. 

How would you improve on the plugin hosts design?

**Comment by GOATS2K** [comment]
> I would propose to move IServiceProxy and IHostServiceProxy to Coral.PluginBase and make plugins depend only on Coral.PluginBase. This would ensure that plugins are given access only to the most essential interfaces and types they need to provide a working plugin implementation.

Would you like to create a PR for this?

**Comment by CaffeinatedCoder** [comment]
Actually I found your GitHub repo and the Coral project by reading your really good article [Plugins in .NET: Creating the plugin host](https://nectarine.sh/posts/plugins-in-dotnet-host/). 
Your codebase inspired me to derive a similar system based on your ideas. I will soon release some essential packages as well.
I could develop a PR for your codebase of course. I also developed a Roslyn analyzer that takes care, that assemblies only ever provide exactly one plugin implementation only, and that the implementations are declared as public classes with a public, parameter-less constructor as well.
